### PR TITLE
Add a test for using a PSK with QUIC

### PR DESCRIPTION
**Pull request by mattcaswell**
Check that we can set and use a PSK when establishing a QUIC connection.

Fixes https://github.com/openssl/project/issues/83

This also fixes a bug in the `SSL_CIPHER_find()` function which crashed when passed a QUIC SSL object.

### CODE REVIEW

File: test/helpers/ssltestlib.c [hunk 1]
```diff
@@ -1247,3 +1247,41 @@ void shutdown_ssl_connection(SSL *serverssl, SSL *clientssl)
...
+
+SSL_SESSION *create_a_psk(SSL *ssl, size_t mdsize)
+{
+    const SSL_CIPHER *cipher = NULL;
+    const unsigned char key[] = {
+        0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a,
+        0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15,
+        0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x20,
+        0x21, 0x22, 0x23, 0x24, 0x25, 0x26, 0x27, 0x28, 0x29, 0x2a, 0x2b,
+        0x2c, 0x2d, 0x2e, 0x2f /* SHA384_DIGEST_LENGTH bytes */
```
Comment (paulidale) [inline test/helpers/ssltestlib.c:1259]: Consider moving the length in the declaration?  That is:
```c
const unsigned char key[SHA384_DIGEST_LENGTH] = {
```
and drop this comment.

The compiler can then do some checking for us.

File: ssl/quic/quic_local.h [hunk 2]
```diff
@@ -321,7 +321,7 @@ const SSL_METHOD *func_name(void)  \
                 NULL /* dispatch_alert */, \
                 ossl_quic_ctrl, \
                 ossl_quic_ctx_ctrl, \
-                NULL /* get_cipher_by_char */, \
+                ssl3_get_cipher_by_char, \
```
Comment (hlandau) [inline ssl/quic/quic_local.h:324]: I suppose this is needed to make SSL_CIPHER_find work.

However we only support a restricted set of ciphersuites for QUIC &mdash; specifically AES-128-GCM, AES-256-GCM and ChaCha20-Poly1305. Do we need to use a custom implementation that restricts what can be retrieved here?

Adding a test that e.g. CCM cannot be retrieved via SSL_CIPHER_find might be nice while we are here.

**Review by hlandau — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
@hlandau - feedback addressed. Please take another look.

**Review by paulidale — APPROVED** [review]

**Review by hlandau — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you.

**Pull request closed without merging**